//
// Copyright (c) 2015 - 2022, Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
//

// TODO: ADD OVERVIEW COMMENT

syntax = "proto3";

message Report {
  string geopm_version = 1;
  repeated HostReport hosts = 2;
}

message TimeSpec {
  uint64 sec = 1;
  uint64 usec = 2;
}

message HostReport {
  string hostname = 1;
  TimeSpec begin = 2;
  TimeSpec end = 3;
  double duration = 4;
  repeated Stats stats = 5;
}

message Stats {
  string name = 1;
  double first = 2;
  double last = 3;
  int64 count = 4;
  double min = 5;
  double max = 6;
  double mean = 7;
  double std = 8;
  double skew = 9;
  double kurt = 10;
  double slope = 11;
}

message Policy {
  string agent = 1;
  double period = 2;
  string profile = 3;
  map<string, double> params = 4;
}

service GEOPMRuntime {
  // Agent determines which statistics will be included in the report
  rpc SetPolicy(Policy) returns(TimeSpec);
  // Get statistics about sampled data since last call to GetReport()
  rpc GetReport(TimeSpec) returns(Report);
  // Forward calls to SetPolicy() and GetReport() to children.  The
  // host sections include reports from children, and Policy messages are
  // forwarded.
  rpc AddChildNode(Url) returns(TimeSpec)
  rpc RemoveChildNode(Url) returns(TimeSpec)
}
